messaging/src/main/java/org/axonframework/eventhandling/EventProcessor.java

```diff
@@ -59,4 +59,14 @@
      * Stop processing events.
      */
     void shutDown();
+
+    /**
+     * Returns the priority in the shutdown cycle of this processor
+     * A higher value means it is shutdown earlier
+     *
+     * @return the shutdownPriority of this event processor
+     */
+    default int shutDownPriority() {
```

Comment:
What about introducing a `phase` (similar to one in `ModuleConfiguration`)? It could give us greater flexibility in future.

Reply:
This was the first thing I thought of.

But in the mean time I have heard you guys have bigger plans for the shutdown order so don't feel obliged to use this pull request.

Reply:
Hi,
I realize I wasn't being very explicit when mentioning the "bigger plans". One of them was actually to use phases for the lifecycle callbacks.

Tracking processors are "edge components", as they are components pulling external data into the application. This makes them candidates to shutdown very early.

So instead of configuring a shutdown priority on components themselves, it would be nicer to have this configured (only) in the Configuration API components.  Shutdown callback registration should be accompanied with a phase (defaulting to 0).

When executing the shutdown, element with the same phase may be shut down concurrently, but elements with a later phase (higher number) may only be shut down when the earlier phases have completed the shutdown process. Spring uses a `Runnable` parameter that shutdown commands can use to indicate they're ready. We could use that as well.

Hope this gives a bit of insight into what we were planning to do. This all most likely takes more than a single PR to get working. I could lay the groundwork and commit that to a WIP-branch, and then we could merge this PR and whatever is coming into that. Would you be up for that, @nurk? If not, that's ok too. Don't worry.

Reply:
I understand.

This feels like a bit too much work.  I make these PR's in my free time hence they are all small and simple.
Feel free to close the pull request.